File: JavaAlgorithm/src/main/java/algo/tzashinorpu/DpPlan/LBLD/coinChange_322.java
package algo.tzashinorpu.DpPlan.LBLD;

import java.util.Arrays;

public class coinChange_322 {
    public int coinChange(int[] coins, int amount) {
        int INF = amount + 1;
        int[] dp = new int[INF];
        Arrays.fill(dp, INF);
        dp[0] = 0;
        for (int i = 1; i <= amount; i++) {
            for (var coin : coins) {
                if (i - coin < 0) continue;
                dp[i] = Math.min(dp[i], 1 + dp[i - coin]);
            }
        }
        return dp[amount] == INF ? -1 : dp[amount];
    }
}

File: Clion/ZeroTrac/DP/coinChange_322.cpp
//
// Created by TzashiNorpu on 3/13/2023.
//
#include "dp.h"

using namespace ZeroTrac;

int Solution::coinChange(vector<int> &coins, int amount) {
  vector<int> dp(amount + 1, amount + 1);
  dp[0] = 0;
  for (int i = 1; i <= amount; ++i) {
    for (int coin: coins) {
      if (i - coin < 0) continue;
      dp[i] = min(dp[i], 1 + dp[i - coin]);
    }
  }
  return (dp[amount] == amount + 1) ? -1 : dp[amount];
}